File: Examples/erlang/algorithms.erl
# Erlang Algorithms & Data Structures
# Time Warp Studio - Classic algorithms in functional style

-module(algorithms).
-export([main/0]).

% ── Sorting ───────────────────────────────────────────────────────────────────

quicksort([]) -> [];
quicksort([Pivot|Rest]) ->
    Less    = [X || X <- Rest, X =< Pivot],
    Greater = [X || X <- Rest, X > Pivot],
    quicksort(Less) ++ [Pivot] ++ quicksort(Greater).

mergesort([])  -> [];
mergesort([X]) -> [X];
mergesort(List) ->
    Mid = length(List) div 2,
    {Left, Right} = lists:split(Mid, List),
    merge(mergesort(Left), mergesort(Right)).

merge([], R) -> R;
merge(L, []) -> L;
merge([H1|T1], [H2|T2]) when H1 =< H2 -> [H1 | merge(T1, [H2|T2])];
merge([H1|T1], [H2|T2])               -> [H2 | merge([H1|T1], T2)].

% ── Searching ─────────────────────────────────────────────────────────────────

binary_search(List, Target) ->
    binary_search(List, Target, 0, length(List) - 1).
binary_search(_, _, Low, High) when Low > High -> not_found;
binary_search(List, Target, Low, High) ->
    Mid = (Low + High) div 2,
    Val = lists:nth(Mid + 1, List),
    if
        Val =:= Target -> {found, Mid};
        Val < Target   -> binary_search(List, Target, Mid + 1, High);
        true           -> binary_search(List, Target, Low, Mid - 1)
    end.

% ── Map / Dict operations ─────────────────────────────────────────────────────

% Word frequency count using a map
word_count(Words) ->
    lists:foldl(
        fun(W, Map) ->
            Count = maps:get(W, Map, 0),
            maps:put(W, Count + 1, Map)
        end,
        #{},
        Words
    ).

% ── Graph traversal (adjacency list) ─────────────────────────────────────────

% BFS using a queue (list as queue)
bfs(Graph, Start) ->
    bfs(Graph, [Start], [Start], []).
bfs(_, [], Visited, _) -> lists:reverse(Visited);
bfs(Graph, [H|Queue], Visited, _) ->
    Neighbors = maps:get(H, Graph, []),
    NewNeighbors = [N || N <- Neighbors, not lists:member(N, Visited)],
    bfs(Graph, Queue ++ NewNeighbors, Visited ++ NewNeighbors, []).

% ── Prime sieve ───────────────────────────────────────────────────────────────

sieve(Max) -> sieve(lists:seq(2, Max), []).
sieve([], Primes) -> lists:reverse(Primes);
sieve([H|T], Primes) ->
    Filtered = [X || X <- T, X rem H =/= 0],
    sieve(Filtered, [H | Primes]).

% ── Main ──────────────────────────────────────────────────────────────────────

main() ->
    io:format("=== Erlang Algorithms & Data Structures ===~n~n"),

    %% Sorting
    io:format("Sorting:~n"),
    Unsorted = [38, 27, 43, 3, 9, 82, 10, 1, 55, 17],
    io:format("  Input:     ~w~n", [Unsorted]),
    io:format("  Quicksort: ~w~n", [quicksort(Unsorted)]),
    io:format("  Mergesort: ~w~n", [mergesort(Unsorted)]),

    %% Binary search
    io:format("~nBinary Search:~n"),
    Sorted = lists:sort(Unsorted),
    io:format("  Sorted: ~w~n", [Sorted]),
    lists:foreach(fun(T) ->
        Res = binary_search(Sorted, T),
        io:format("  search(~w) -> ~w~n", [T, Res])
    end, [3, 55, 42]),

    %% Word frequency
    io:format("~nWord Frequency:~n"),
    Words = [the, cat, sat, on, the, mat, the, cat, sat],
    Freq = word_count(Words),
    maps:foreach(fun(W, C) ->
        io:format("  ~w: ~w~n", [W, C])
    end, Freq),

    %% BFS graph traversal
    io:format("~nBFS Graph Traversal:~n"),
    Graph = #{
        a => [b, c],
        b => [a, d, e],
        c => [a, f],
        d => [b],
        e => [b, f],
        f => [c, e]
    },
    Order = bfs(Graph, a),
    io:format("  BFS from 'a': ~w~n", [Order]),

    %% Prime sieve
    io:format("~nPrimes up to 50 (Sieve of Eratosthenes):~n"),
    Primes = sieve(50),
    io:format("  ~w~n", [Primes]),
    io:format("  Count: ~w~n", [length(Primes)]),

    io:format("~nDone.~n").

File: Examples/python/algorithms.py
# Classic algorithms in Python
# Demonstrates sorting, searching, and mathematical algorithms

import math

# --- Sorting algorithms ---

def bubble_sort(lst):
    arr = list(lst)
    n = len(arr)
    for i in range(n):
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr


def merge_sort(lst):
    if len(lst) <= 1:
        return list(lst)
    mid = len(lst) // 2
    left = merge_sort(lst[:mid])
    right = merge_sort(lst[mid:])
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i]); i += 1
        else:
            result.append(right[j]); j += 1
    return result + left[i:] + right[j:]


def binary_search(arr, target):
    lo, hi = 0, len(arr) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1


# --- Mathematical algorithms ---

def is_prime(n):
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False
    return True


def sieve_of_eratosthenes(limit):
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False
    for i in range(2, int(math.sqrt(limit)) + 1):
        if sieve[i]:
            for j in range(i * i, limit + 1, i):
                sieve[j] = False
    return [i for i, v in enumerate(sieve) if v]


def gcd(a, b):
    while b:
        a, b = b, a % b
    return a


def fibonacci(n):
    a, b = 0, 1
    result = []
    for _ in range(n):
        result.append(a)
        a, b = b, a + b
    return result


# --- Run demonstrations ---

print("=== Sorting ===")
data = [64, 34, 25, 12, 22, 11, 90]
print(f"Original:   {data}")
print(f"Bubble:     {bubble_sort(data)}")
print(f"Merge:      {merge_sort(data)}")

print("\n=== Binary Search ===")
sorted_data = merge_sort(data)
for target in [25, 90, 5]:
    idx = binary_search(sorted_data, target)
    if idx >= 0:
        print(f"Found {target} at index {idx}")
    else:
        print(f"{target} not found")

print("\n=== Primes (Sieve, up to 50) ===")
primes = sieve_of_eratosthenes(50)
print(primes)

print("\n=== Fibonacci (first 12) ===")
print(fibonacci(12))

print("\n=== GCD Examples ===")
pairs = [(48, 18), (100, 75), (17, 5)]
for a, b in pairs:
    print(f"gcd({a}, {b}) = {gcd(a, b)}")
